### ISSUE TITLE
DatePicker: Value prop not set when using onSelectDate callback

### ISSUE THREAD

**Issue by WesselKuipers** [OfficeDev/office-ui-fabric-react#1274]
### Bug Report

- __Package version(s)__: 2.4.0
- __Browser and OS versions__: Chrome 56.0.2924.87 (64-bit)

#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? (Yes, No)
No

Requested priority: (Blocking, High, Normal, Low) 
Normal

#### Describe the issue:

The DatePicker component has an onSelectDate callback which is issued whenever the user selects a new date. If onSelectDate is implemented, the DatePicker no longer seems to assign the new value in its own props. This causes the DatePicker to not display the date the user selected.

It's possible to work around this by manually assigning the value, but this behaviour seems unintended regardless.

Another strange issue surfaced while I made the codepen repro: If I have two DatePicker components, the second of which has onSelectDate implemented, the date of the first DatePicker is no longer shown after picking a date on the second DatePicker.

#### Actual behavior:

DatePicker doesn't show selected date when onSelectDate is implemented.

#### Expected behavior:

DatePicker should display the selected date, even if onSelectDate is implemented.

### If applicable, please provide a codepen repro:

http://codepen.io/WesselKuipers/pen/zZEGpm

**Comment by mdahamiwal** [comment]
Its more of an implementation issue:

```
<Fabric.DatePicker required={true} />
<span>Datepicker with onSelectDate callback:</span>
<Fabric.DatePicker required={true} onSelectDate={newDate => {console.log(newDate); this.setState({date: newDate})}} /> 
```
both DatePickers have no initial value (date) due to which when this.setState() is called in second date picker it causes re-render of container component and cleans-up user selections in DatePickers. This is well demonstrated in examples provided on demo site or try http://codepen.io/mdahamiwal/pen/vxpvVW . Hope it helps.

**Comment by WesselKuipers** [comment]
It strikes me as a bit strange or at least inconsistent that you have to keep track of the internal value yourself when you don't have to do this with components like TextField.

The main issue I was trying to report was the fact that when you implement the onSelectDate callback to get notified when a user picks a date, the DatePicker component suddenly doesn't display the date that was just picked.

Is there a reason why this behaviour is inconsistent compared to when you implement onChange in TextField?

**Comment by mdahamiwal** [comment]
Yes, not retaining selectedDate seems like a bug at 
https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.tsx#L123-L128

**Comment by johannao76** [comment]
What ends up happening here is that the setState call on the parent component triggers an update in the DatePicker, which now needs to update with the new props. However, the call to componentWillReceiveProps during update ends up incorrectly resetting the this.state.selectedDate to the old prop value. A potential fix for this is to check the old and new date values in componentWillReceiveProps and only set the date to the new value if it has changed. This way, we avoid overwriting the selectedDate in the case when it is changing due to a state change as opposed to a legit prop change.

**Comment by WesselKuipers** [comment]
Looks like the issue has been resolved now, thanks! 👍

**Comment by johannao76** [comment]
Thanks for closing it, forgot to do that!!

Sent from my iPhone

On Apr 19, 2017, at 3:30 AM, Wessel Kuipers <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #1274<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOfficeDev%2Foffice-ui-fabric-react%2Fissues%2F1274&data=02%7C01%7Cjohannao%40microsoft.com%7C2d1f0ba16a024e2da75108d4870f1153%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636281946205724050&sdata=ykghwV1pxj2z%2FF8BSg%2BB9DaImzqwERW0qhpuNMhDrfg%3D&reserved=0>.
